core/src/main/scala/kafka/server/BrokerServer.scala

```diff
@@ -624,23 +624,19 @@ class BrokerServer(
     }
   }
 
-  private def createPartitionMetadataClient(): PartitionMetadataClient = {
-    // This is a no-op implementation of PartitionMetadataClient. It always returns -1 as the latest offset for any
-    // requested topic partition.
-    // TODO: KAFKA-19800: Implement a real PartitionMetadataClient that can fetch latest offsets via InterBrokerSendThread.
-    new PartitionMetadataClient {
-      override def listLatestOffsets(topicPartitions: util.Set[TopicPartition]
-                                    ): util.Map[TopicPartition, util.concurrent.CompletableFuture[java.lang.Long]] = {
-        topicPartitions.asScala
-          .map { tp =>
-            tp -> CompletableFuture.completedFuture(java.lang.Long.valueOf(-1L))
-          }
-          .toMap
-          .asJava
-      }
-
-      override def close(): Unit = {}
-    }
+  private def createNetworkPartitionMetadataClient(metadataCache: MetadataCache): PartitionMetadataClient = {
+    new NetworkPartitionMetadataClient(
+      metadataCache,
+      NetworkUtils.buildNetworkClient(
+        "NetworkPartitionMetadataClient",
+        config,
+        metrics,
+        Time.SYSTEM,
+        new LogContext(s"[NetworkPartitionMetadataClient broker=${config.brokerId}]")
+      ),
```

Comment:
We need to decide if we want the `PartitionMetadataClient` to be lazily loaded, if yes then we should pass a provider which can fetch the NetworkClient.